### ISSUE TITLE
Text created in the empty space cannot be moved by API

### ISSUE THREAD

**Issue by Fredosixx** [SketchUp/api-issue-tracker#854]
A small bug which I discovered accidentally.

1. Create a text _in the empty space_ via the Text Label tool
2. The text entity has actually no `point` and `vector` defined (i.e. `etext.point = nil` and `etext.vector = nil`
3. From there, the method `#transform_entities` will have no effect. You cannot move the text label via the API.
4. Strangely, you cannot event change the `point` and `vector` attributes: `etext.point = ORIGIN` will not return an error, but right after, `etext.point` will still be `nil`.
5. Finally, only the native Move tool is able to move the text label

**Comment by DanRathbun** [comment]
I know I have reported this years ago (perhaps before the public GitHub Issue Tracker.)

The reason that a 3D transform does **not** apply, is because this is 2D **screen space text**.

Ie ... it is the same "special" kind of object created with the [`Sketchup::Model#add_note`](https://ruby.sketchup.com/Sketchup/Model.html#add_note-instance_method) method.

The only way to move a screen note object is to recreate it and erase the old one.
```ruby
note = ents.grep(Sketchup::Text).find { |t|
  t.point.nil? && t.vector.nil? && !t.has_leader? && t.leader_type == ALeaderNone
}
# Where the 2D screen positions are float percentages from top left of screen:
new_note = model.add_note(note.text, new_x_screen_percentage, new_y_screen_percentage)
# Delete the original:
note.erase!
```

**Comment by DanRathbun** [comment]
> _I know I have reported this years ago (perhaps before the public GitHub Issue Tracker.)_

No it was here. I reported it as an enhancement / new subclass just for view locked notes:
[Issue 207 : View notes could be a subclass of Sketchup::Text ](https://github.com/SketchUp/api-issue-tracker/issues/207)

This came about from a public forum discussion on all sorts of API issues. So it's buried in the middle of the thread:
https://forums.sketchup.com/t/ruby-api-documentation-errors-suggestions/10958/7

----
RELATED:

[Issue 556 : Screen text notes do not display changes (using set_text) until view is redrawn](https://github.com/SketchUp/api-issue-tracker/issues/556)

**Comment by Fredosixx** [comment]
The method `model#add_note` returns a `Sketchup::Text` object. However, like a text label placed manually in the empty space, the `point` attribute is `nil`, the `bounds.min' is `Point3d(1e+30, 1e+30, 1e+30)' and the `bounds.max` is `Point3d(-1e+30, -1e+30, -1e+30)`.  And there is no method to get the screen percentage.

So, the suggested workaround (recreate a note) won't work, because there is no way to find out where the text is located on the screen.

Note also that there is another bug concerning model notes, since
```ruby
tx = Sketchup.active_model.add_note('foo', 0.5, 0.5)
tx.set_text 'bar'
```
will create the note, but not update the text visually (however `tx.text` will return 'bar', meaning that the text is updated internally, but not on the screen).

**Comment by DanRathbun** [comment]
> _And there is no method to get the screen percentage._

Yes, I said this and suggested an `#anchor` getter method almost 8 years ago. Again, see Issue [207](https://github.com/SketchUp/api-issue-tracker/issues/207).

> _Note also that there is another bug concerning model notes, since_ ...
> [_snip_] 
>... _will create the note, but not update the text visually_ ...

I linked to this bug's Issue report ([556](https://github.com/SketchUp/api-issue-tracker/issues/556)) above under "RELATED".

> _So, the suggested workaround (recreate a note) won't work, because there is no way to find out where the text is located on the screen._

**On the contrary.** If you know where you want the note, regardless of where it is now, then discarding any current location and setting a completely new location **will work**. I have done it.

**Comment by DanRathbun** [comment]
## Bounds

>... _the `bounds.min` is `Point3d(1e+30, 1e+30, 1e+30)` and the `bounds.max` is `Point3d(-1e+30, -1e+30, -1e+30)`._

Using 3D bounds will always fail for a 2D screen space object.

Because the font size can be changed, we would need to get a [`Geom::Bounds2d`](https://ruby.sketchup.com/Geom/Bounds2d.html) object for screen notes.

**Comment by Fredosixx** [comment]
My use case is a plugin that mimic the Move tool:
1) A model contains a Text / Note element
2) The plugin allows the user to select it (by API, as it is detected by `PickHelper`)
3) The plugin wants to perform a move from an origin to a target

Since there is no way to know its current position, and it does not react to `transform_entities` it is impossible to move it by API.

**Comment by DanRathbun** [comment]
I am not in any way against fixing the issues with screen notes. And, I don't know why you want to mimic the Move tool in this case, but this is not important for me to know.

If you are using [`PickHelper`](https://ruby.sketchup.com/Sketchup/PickHelper.html) in your tool, then wouldn't you have the X,Y screen coordinates from which to start the move ? If not, then you'd need to pick some convenient screen position to start. Perhaps the center of the view.

So, instead of using a transform to drag the actual object around, ... save it's text, delete it, and then use [`View#draw_text`](https://ruby.sketchup.com/Sketchup/View.html#draw_text-instance_method) to show a facsimile of the view note attached to the cursor during the [`onMouseMove`](https://ruby.sketchup.com/Sketchup/Tool.html#onMouseMove-instance_method) callback. 

Then when the user clicks to place the text, recreate where they clicked.

And yes there are other limitations such as not knowing the font face and size, etc.

**Comment by DanRathbun** [comment]
> _And yes there are other limitations such as not knowing the font face and size, etc._

Actually I meant limitations in the Ruby API.

You might be able use the Live C API to get and set the screen position, (as well as the color and the font.)

- [`SUTextGetScreenPosition()`](https://extensions.sketchup.com/developers/sketchup_c_api/sketchup/struct_s_u_text_ref.html#ad79f59deb1d198e71b81847d244ebb22)
- [`SUTextSetScreenPosition()`](https://extensions.sketchup.com/developers/sketchup_c_api/sketchup/struct_s_u_text_ref.html#abcd506e70d1dbc94c013ccd8172e2942)

@thomthom Would it be safe to set the position using the C API yet ?

Fredo, at the very least you can **get** the position.

**Comment by Fredosixx** [comment]
Very interesting...

I am not familiar with the live C API however. Is there a method to transform a Ruby object (the Text object) into the binary handle `SUTextRef` for the C API?

**Comment by DanRathbun** [comment]
> Is there a method to transform a Ruby object (the Text object) into the binary handle `SUTextRef` for the C API?

Yes. [`SUEntityFromRuby`](https://extensions.sketchup.com/developers/sketchup_c_api/sketchup/ruby__api_8h.html#a51510ff046c085ed40f57611b0558258) and then downcast using [`SUTextFromEntity`](https://extensions.sketchup.com/developers/sketchup_c_api/sketchup/struct_s_u_text_ref.html#aeb1c46f25df0eb2ac2bc74e62d6dc2c5).

Use the C Ruby function `rb_float_new` to convert the C double to a Ruby `Float` when returning the values to the Ruby side.
